File: solutions/java/need-for-speed/1/src/main/java/NeedForSpeed.java
class NeedForSpeed {
    public static final int MAX_BATTERY = 100;
    
    private int speed;
    private int batteryDrain;
    private int batteryLevel = MAX_BATTERY;
    private int metersDriven;
    
    NeedForSpeed(int speed, int batteryDrain) {
        this.speed = speed;
        this.batteryDrain = batteryDrain;
    }

    public boolean batteryDrained() {
        return batteryDrain > batteryLevel;
    }

    public int distanceDriven() {
        return metersDriven;
    }

    public void drive() {
        if (batteryDrained()) {
            return;
        }
        
        metersDriven += speed;
        batteryLevel -= batteryDrain;
    }

    public int getBatteryDrain() {
        return batteryDrain;
    }

    public int getSpeed() {
        return speed;
    }

    public static NeedForSpeed nitro() {
        return new NeedForSpeed(50, 4);
    }
}

class RaceTrack {
    private int distance;
    
    RaceTrack(int distance) {
        this.distance = distance;
    }

    public boolean canFinishRace(NeedForSpeed car) {
        int distanceCovered = (NeedForSpeed.MAX_BATTERY / car.getBatteryDrain()) * car.getSpeed();
        
        return distanceCovered >= distance;
    }
}

File: solutions/csharp/need-for-speed/1/NeedForSpeed.cs
class RemoteControlCar
{
    private int speed;
    private int batteryDrain;
    public int batteryPercentage;
    private int distanceDriven;

    public RemoteControlCar(int speed, int batteryDrain)
    {
        this.speed = speed;
        this.batteryDrain = batteryDrain;
        batteryPercentage = 100;
        distanceDriven = 0;
    }

    public bool BatteryDrained() => batteryPercentage < batteryDrain;

    public int DistanceDriven() => distanceDriven;

    public void Drive()
    {
        if (BatteryDrained()) 
        {
            return;
        }

        distanceDriven += speed;
        batteryPercentage -= batteryDrain;
    }

    public static RemoteControlCar Nitro() => new RemoteControlCar(50, 4);
}

class RaceTrack
{
    private int distance;

    public RaceTrack(int distance)
    {
        this.distance = distance;
    }

    public bool TryFinishTrack(RemoteControlCar car)
    {
        while (!car.BatteryDrained())
        {
            car.Drive();
        }

        return car.DistanceDriven() >= distance;
    }
}
